### PR TITLE
Fix user defined preprocess function missing prediction issue

### PR DESCRIPTION
**Pull request by zane-neo**
### Description
When users use self defined preprocess scripts to process inputs, there'll be missing predictions for all inputs in the list except the first one.
 
### Issues Resolved
https://github.com/opensearch-project/ml-commons/issues/2417
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java [hunk 1]
```diff
@@ -99,6 +102,15 @@ private Tuple<Integer, Integer> calculateChunkSize(TextDocsInputDataSet textDocs
                 return Tuple.tuple(textDocsLength / stepSize + 1, stepSize);
             }
         } else {
+            Optional<ConnectorAction> predictAction = getConnector().findPredictAction();
```
Comment (dhrubo-os) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java:105]: Can we have unit test for this section?

File: common/src/main/java/org/opensearch/ml/common/connector/MLPreProcessFunction.java [hunk 2]
```diff
@@ -22,6 +23,7 @@ public class MLPreProcessFunction {
     public static final String TEXT_DOCS_TO_COHERE_EMBEDDING_INPUT = "connector.pre_process.cohere.embedding";
     public static final String TEXT_DOCS_TO_OPENAI_EMBEDDING_INPUT = "connector.pre_process.openai.embedding";
     public static final String TEXT_DOCS_TO_BEDROCK_EMBEDDING_INPUT = "connector.pre_process.bedrock.embedding";
+    public static final String TEXT_IMAGE_TO_BEDROCK_EMBEDDING_INPUT = "connector.pre_process.multimodal.embedding";
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/connector/MLPreProcessFunction.java:26]: Can we raise a separate PR for this functionality only.

**Comment by ylwu-amzn** [comment]
Suggest not merge multiple things in one PR. That makes it hard to rollback/backport

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java [hunk 3]
```diff
@@ -99,6 +102,15 @@ private Tuple<Integer, Integer> calculateChunkSize(TextDocsInputDataSet textDocs
...
+            if (predictAction.isEmpty()) {
+                throw new IllegalArgumentException("no predict action found");
+            }
+            String preProcessFunction = predictAction.get().getPreProcessFunction();
+            if (preProcessFunction != null && !MLPreProcessFunction.contains(preProcessFunction)) {
+                // user defined preprocess script, this case, the chunk size is always equals to text docs length.
+                return Tuple.tuple(textDocsLength, 1);
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java:112]: Is this understanding correct ?
1. `textDocsLength` means how many chunks
2. `1` means step or chunk size ?

If correct, why hard code the chunk size as 1 ?  This issue https://github.com/opensearch-project/ml-commons/issues/2417 is an example, it's not the only case. For example some user may process two documents in one chunk.

Comment (ylwu-amzn) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java:112]: https://github.com/opensearch-project/ml-commons/blob/2.x/docs/remote_inference_blueprints/bedrock_connector_titan_multimodal_embedding_blueprint.md

Comment (zane-neo) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java:112]: textDocsLength is chunks, 1 means step. When user process two documents in one chunk, user has to specify the `input_docs_processed_step_size`, in this case, this won't enter this branch. Multi-modal is a case that user needs to handle two documents in one chunk, I've tested this case and it works well.

**Review by dhrubo-os — APPROVED** [review]

**Review by xinyual — APPROVED** [review]

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java (hunk 3 above)
Comment (ylwu-amzn) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java:112]: Thanks for testing multi-modal case. 
It's possible that step size > 1 and no `input_docs_processed_step_size`. For example, before async http client, user can confiture pre_process function to read three documents and construct `"inputs": [doc1, doc2, doc3]`, when response comes back, will move to next 3 docs

Comment (zane-neo) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java:112]: This is a rare case, and we have workaround which is to add `input_docs_processed_step_size` configuration.

Comment (ylwu-amzn) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java:112]: Thanks, based on our discussion, this sounds good

**Pull request merged**